### DISCUSSION TITLE
Abnormal operation on Teensy 4.1 dedicated micro sdcard

### DISCUSSION THREAD

**Discussion by LeandroFernandes88** [micropython/micropython#15140] [m.iMX RT / Teensy 4.x]
hello everybody.
I noticed abnormal behavior on the Teensy 4.1 sdcard
Firstly, I used the internal micro SDcard slot with the microcontroller connected to the PC's power supply, everything worked normally.
So I connected a battery and 5v regulator to the 5v input pin, so the internal sdcard slot stopped working and I could only access it with an external micro sdcard board.

**Comment by robert-hh** [comment]
Looking at the schematics there is no reason for a difference. VUSB and Vin are connected at the board. How do tell whether the SD card works at boot or not? Does your code rely on the SD card being mounted at boot or does it try top mount the SD card later?

**Reply by robert-hh** [reply]
I just made a test and cannot replicate the behavior. When the SD card is inserted at power-on by applying 5V to Vin, the SD card is mounted and **main.py from the SD card** is executed.

**Reply by LeandroFernandes88** [reply]
OK. I will try again to run the code on battery.
As I prepared two different codes, to read and write to the external slot and the integrated slot, I'm going to check if there could have been a silly error somewhere in the code.
Thanks for all the help

**Reply by LeandroFernandes88** [reply]
Hi [robert-hh](https://github.com/robert-hh)
I made a video showing the problem, I don't know if I can upload it to the forum.

In my code, I create a folder on the sdcard and then turn on the green light on both the RP2040 and the Teensy4.1, however on the teensy nothing happens, only when I am using the external sdscard.

One question, in Thonny, just when you insert the card the sdcard is mounted.
In the battery I need to add some code like:
os.mount('/sdcard') ?

PS: sorry for the background noise and poor english.


https://github.com/micropython/micropython/assets/125308622/e038f417-4b4b-4559-b4cc-c1091a71215b

**Reply by LeandroFernandes88** [reply]
> Does your code rely on the SD card being mounted at boot or does it try top mount the SD card later?

I didn't mentioned the sdcard being mounted.
It mounts automatically using Thonny, so I imagined that it will do the same on battery, or not?

**Reply by LeandroFernandes88** [reply]
I would like to mention some tests I also did using the Teensy4.1's internal slot (on the battery) and with an external slot and the sdcard.py library, probably this one:
https://github.com/micropython/micropython-lib/blob/master/micropython/drivers/storage/sdcard/sdcard.py

SDcard performance times: dedicated sdcard vs external sdcard
Create a folder: 3ms x **27.5ms**
save .bin file size 32kb: 14ms x **90 ms**

It may have been how to set the external sd card mount?
(I did absolutely nothing internally, I just gave the command to save in: '/sdcard/...)

```
# Configure sdcard Pin --  3.3V(OUT) red / GND black
SD_MOSI = Pin('D11', Pin.OUT) #D11 - Pin.board.MOSI ou SPIO TX - yellow
SD_SCK = Pin('D13', Pin.OUT)  #D13 - Pin.board.SCK     SPIO SCK - green
SD_CS = Pin('D10', Pin.OUT)   #D10 - Pin.board.D0      SPIO CSn - blue
SD_MISO = Pin('D12', Pin.IN)  #D12 - Pin.board.MISO    SPIO RX - white    
#
spi = SPI(0,
          baudrate=10_000_000,
          polarity=0,
          phase=0,
          bits=8,
          firstbit=SPI.MSB)
#
cs = SD_CS
#
sd = sdcard.SDCard(spi, cs, baudrate=10_000_000)
vfs = os.VfsFat(sd)
os.mount(vfs, '/sd')   # just external sdcard
```

**Reply by robert-hh** [reply]
The internal SDCARD driver is faster because:

- it uses quad mode for transfer (4bit), and
- it is all coded in C.

The SDCard is mounted at soft reset and hard reset. Thonny performs frequent soft resets, typically at each user action. Therefore it seems as if the SDCard is automatically mounted. When using an external power or some other terminal than Thonny, you will not see the automatic mount. Unfortunately the card detect switch is not connected at the Teensy Hardware, and the sdcard.present() function switches from False to True only once after power up, when you insert a card the first time.
